open_spiel/python/games/dynamic_routing/mean_field_routing_game.py

```diff
@@ -0,0 +1,547 @@
+        """Initiliaze the game.
+
+        Args:
+            params: game parameters.
+            network: the network of the game.
+            od_demand: a list of the vehicle. Their origin and their
+                destination should be road sections of the game. The number of
+                vehicles in the list set the num_player attribute.
+            max_num_time_step: set the max_game_length attribute.
+            perform_sanity_checks: if true, sanity checks are done during the
+                game, should be set to false to faster the game.
+        """
+        self.network = network if network else _DEFAULT_NETWORK
+        self._od_demand = od_demand if od_demand else _DEFAULT_DEMAND
+        self.network.check_list_of_od_demand_is_correct(self._od_demand)
+        self.perform_sanity_checks = perform_sanity_checks
+        game_info = pyspiel.GameInfo(
+            num_distinct_actions=self.network.num_links() + INDEX_FIRST_ACTION,
+            max_chance_outcomes=max(2, len(self._od_demand)),
+            num_players=1,
+            min_utility=-max_num_time_step-1,
+            max_utility=0,
+            max_game_length=max_num_time_step)
+        super().__init__(_GAME_TYPE, game_info, params if params else {})
+
+    def new_initial_state(self) -> "MeanFieldRoutingGameState":
+        """Returns the state corresponding to the start of a game."""
+        return MeanFieldRoutingGameState(self, self._od_demand)
+
+    def make_py_observer(self, iig_obs_type=None, params=None):
+        """Returns a NetworkObserver object used for observing game state."""
+        return NetworkObserver(
+            self.network.num_links(), self.max_game_length())
+
+    def max_chance_nodes_in_history(self):
+        """Maximun chance nodes in game history."""
+        return self.max_game_length() + 1
+
+
+class MeanFieldRoutingGameState(pyspiel.State):
+    """State of the DynamicRoutingGame.
+
+    One player is equal to one vehicle.
+    See docstring of the game class and of the file for more information.
+    Attributes:
+      _current_time_step: current time step of the game.
+      _init_distribution: probability at time 0 for the representative player
+        to have the origin, the destination and the departure time given by
+        _od_demand.
+      _is_chance_init: boolean that encodes is the current node is the initial
+        chance node.
+      _is_terminal: boolean that encodes is the game is over.
+      _normed_density_on_vehicle_link: density of vehicle on the link that is
+        used by the representative vehicle. This is given by the mean field
+        distribution.
+      _total_num_vehicle: total number of vehicles as the sum of the _od_demand.
+      _vehicle_at_destination: Set of vehicles that have reached their
+        destinations. When a vehicle has reached its destination but the game
+        is not finished, it cannot do anything.
+      _vehicle_destination: the destination of each vehicle.
```

Comment:
Of the representative vehicle corresponding to this state (once the state is no longer in chance_init mode).